Mathlib/Algebra/Module/Submodule/Lattice.lean

```diff
@@ -223,34 +218,73 @@ instance : InfSet (Submodule R M) :=
       add_mem' := by simp (config := { contextual := true }) [add_mem]
       smul_mem' := by simp (config := { contextual := true }) [smul_mem] }⟩
 
-private theorem sInf_le' {S : Set (Submodule R M)} {p} : p ∈ S → sInf S ≤ p :=
-  Set.biInter_subset_of_mem
-
-private theorem le_sInf' {S : Set (Submodule R M)} {p} : (∀ q ∈ S, p ≤ q) → p ≤ sInf S :=
-  Set.subset_iInter₂
-
 instance : Inf (Submodule R M) :=
   ⟨fun p q ↦
     { carrier := p ∩ q
       zero_mem' := by simp [zero_mem]
       add_mem' := by simp (config := { contextual := true }) [add_mem]
       smul_mem' := by simp (config := { contextual := true }) [smul_mem] }⟩
 
+instance : SupSet (Submodule R M) where
+  sSup S :=
+    { toAddSubmonoid := sSup (toAddSubmonoid '' S)
+      smul_mem' := fun t {m} hm ↦ by
+        simp only [AddSubsemigroup.mem_carrier, AddSubmonoid.mem_toSubsemigroup] at hm ⊢
+        rw [sSup_eq_iSup'] at hm ⊢
+        refine AddSubmonoid.iSup_induction'
+          (C := fun x _ ↦ t • x ∈ ⨆ p : toAddSubmonoid '' S, (p : AddSubmonoid M)) ?_ ?_
+          (fun x y _ _ ↦ ?_) hm
+        · rintro ⟨-, ⟨p : Submodule R M, hp : p ∈ S, rfl⟩⟩ x (hx : x ∈ p)
+          suffices p.toAddSubmonoid ≤ ⨆ q : toAddSubmonoid '' S, (q : AddSubmonoid M) by
+            exact this (smul_mem p t hx)
+          apply le_sSup
+          rw [Subtype.range_coe_subtype]
+          exact ⟨p, hp, rfl⟩
+        · simpa only [smul_zero] using zero_mem _
+        · simp_rw [smul_add]; exact add_mem }
+
+instance : Sup (Submodule R M) where
+  sup p q :=
+    { toAddSubmonoid := p.toAddSubmonoid ⊔ q.toAddSubmonoid
+      smul_mem' := fun t m (hm : m ∈ p.toAddSubmonoid ⊔ q.toAddSubmonoid) ↦ by
+        change t • m ∈ p.toAddSubmonoid ⊔ q.toAddSubmonoid
+        rw [AddSubmonoid.mem_sup] at hm ⊢
+        obtain ⟨y, hy, z, hz, rfl⟩ := hm
+        exact ⟨t • y, p.smul_mem t hy, t • z, q.smul_mem t hz, (smul_add _ _ _).symm⟩ }
```

Comment:
This one is probably a lot more useful than the `SupSet` one.

Reply:
I want to prioritise my actual goal for now but I may return to tinker with this definition again in a few days. If this change has no performance issues then I'd probably push for it.